### PR TITLE
switch impls based on crate features for u8 as an example

### PR DESCRIPTION
**Pull request by KillingSpark**
As a draft towards #93, and #92

This introduces three crate features which enable choosing the default implementations. The implementation is chosen based on the flag with the most hardware compatibility ( => the highest restrictions on memory usage)

    1. NoTable takes precedence over both others
    2. Bytewise takes precedence over Slice16 and is the default if no feature is selected
    3. Slice16 is used only if no other features are selected

### CODE REVIEW

**Comment by KillingSpark** [comment]
@akhilles Any thoughts on this?

File: src/crc128/default.rs [hunk 1]
```diff
@@ -1,11 +1,44 @@
...
+    not(feature = "bytewise-defaults"),
+    feature = "slice16-defaults"
+))]
+impl Implementation for u128 {
+    type Width = u128;
+    type Table = [[u128; 256]; 16];
+}
 
 impl Crc<u128> {
     pub const fn new(algorithm: &'static Algorithm<u128>) -> Self {
```
Comment (akhilles) [inline src/crc128/default.rs:28]: This fails to compile if the crate is imported with `default-features = false`. I think `bytewise` should be used as the default in that case.

I know some people always import dependencies with `default-features = false` even if the crate doesn't actually make use of features.

File: Cargo.toml [hunk 2]
```diff
@@ -14,6 +14,12 @@ keywords = ["crc", "crc16", "crc32", "crc64", "hash"]
 categories = ["algorithms", "no-std"]
 edition = "2018"
 
+[features]
+default = ["slice16-defaults"]
```
Comment (akhilles) [inline Cargo.toml:18]: Can the `default = ["slice16-defaults"]` part be a separate PR? I think it would be good to have a release with the new functionality + feature flags before changing the defaults. I'm still not sure if this warrants a major version bump.

**Comment by akhilles** [comment]
Overall, LGTM. I think this is the right approach.

File: Cargo.toml (hunk 2 above)
Comment (KillingSpark) [reply Cargo.toml:18]: Yes you are right, this should be able to be released as a minor version bump. So if no explicit action is taken users should still get the `bytewise` behavior.

File: Cargo.toml [hunk 3]
```diff
@@ -14,6 +14,11 @@ keywords = ["crc", "crc16", "crc32", "crc64", "hash"]
 categories = ["algorithms", "no-std"]
 edition = "2018"
 
+[features]
+notable-defaults = []
```
Comment (akhilles) [inline Cargo.toml:18]: [nit] On naming, wdyt about "prefer-low-memory" and "prefer-fast" instead for "notable-defaults" and "slice16-defaults" respectively? They might be easier to understand for some users since it maps closer to intent.

File: Cargo.toml [hunk 4]
```diff
@@ -14,6 +14,11 @@ keywords = ["crc", "crc16", "crc32", "crc64", "hash"]
...
+bytewise-defaults = []
```
[the remark this answers is not in the document]
Comment (KillingSpark) [reply Cargo.toml:19]: Consider a crate that chooses slice16 as the feature.

You want to use the crate but only have enough space for the bytewise implementation.

Then you can use that crate, enable the bytewise-defaults feature and still have the performance benefit of a lookup table which is probably faster than the notable implementation even on simple CPU architectures.

File: Cargo.toml (hunk 3 above)
Comment (KillingSpark) [reply Cargo.toml:18]: I'd expect the people that care about any of this to also care about the specifics i.e. `how much memory is "low memory" exactly` at which point they also need to read about the different implementations. To then name the features after these implementations seems logical to me

File: src/crc128.rs [hunk 5]
```diff
@@ -170,9 +169,65 @@ const fn update_slice16(
 
 #[cfg(test)]
 mod test {
-    use crate::{Bytewise, Crc, NoTable, Slice16};
+    use crate::{Bytewise, Crc, Implementation, NoTable, Slice16};
     use crc_catalog::{Algorithm, CRC_82_DARC};
 
+    #[test]
+    fn default_table_size() {
+        let table_size = core::mem::size_of::<<u128 as Implementation>::Table>();
```
[the remark this answers is not in the document]
Comment (akhilles) [reply src/crc128.rs:177]: These invariants can also be enforced using compile-time assertions. E.g:
```rust
// If "notable-defaults" feature is enabled, memory usage is negligible.
#[cfg(feature = "notable-defaults")]
const _: () = assert!(size_of::<Crc<u128>>() <= 8);

// If "bytewise-defaults" feature is enabled, memory usage is less than ~4 KiB.
#[cfg(feature = "bytewise-defaults")]
const _: () = assert!(size_of::<Crc<u128>>() <= 4 * 1024 + 8);
```

Now, if the default implementation uses more memory than the features allow, it'll be a compile error!

File: tests/test_under_all_feature_combinations.sh
Comment (akhilles) [inline tests/test_under_all_feature_combinations.sh:1]: I think the feature combinations should be tested in CI. It shouldn't really add to the runtime: the tests only take ~2 seconds to run, most of the time is spent downloading dependencies. I don't think that's a blocker for getting this PR merged though. I can update the CI after.

File: src/crc128.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (KillingSpark) [reply src/crc128.rs:177]: I'll do the const thing, that's really cool!

I don't see the benefit in testing the public API for size though. This test is explicitly about detecting wrong choices for the default implementation by inspecting the size of the tables.

File: tests/test_under_all_feature_combinations.sh
Comment (KillingSpark) [reply tests/test_under_all_feature_combinations.sh:1]: I was worried because it will have to compile the crate for each combination. But I forgot it will reuse the compiled dependencies so it's not that much. But I'll leave it to you to integrate it into the CI. The script for it should be runnable in the CI :)

File: Cargo.toml (hunk 3 above)
Comment (KillingSpark) [reply Cargo.toml:18]: I'll have another think about the naming

Comment (KillingSpark) [reply Cargo.toml]: I came up with 

* no-table-memory-restrictions
* bytewise-memory-restrictions
* slice16-memory-restrictions

I think this captures the aspect of the additional restrictions one is imposing by activating the features, hopefully also conveying that those are additive.

**Comment by KillingSpark** [comment]
I don't think there is anything left to do, aside from agreeing on the naming for the features

File: Cargo.toml (hunk 3 above)
Comment (akhilles) [reply Cargo.toml]: Ah, that's a good idea! I've also seen "limit" used in this context. Wdyt about using "mem-limit" or "memory-limit" as the suffix? Has the advantage that it's shorter. Sorry for being so picky on the naming :)

Comment (KillingSpark) [reply Cargo.toml]: > Sorry for being so picky on the naming :)

No worries, naming are important and one of the hardest things to get right :)

"Limit" seems like a fine choice!

**Comment by akhilles** [comment]
LGTM, thanks for your patience!
